src/core/controllers/map-controller.js

```diff
@@ -0,0 +1,218 @@
+  let pitch = startPitch;
+  if (deltaScaleY > 0) {
+    // Gradually increase pitch
+    pitch = startPitch + deltaScaleY * (maxPitch - startPitch);
+  } else if (deltaScaleY < 0) {
+    // Gradually decrease pitch
+    pitch = startPitch - deltaScaleY * (minPitch - startPitch);
+  }
+
+  return viewState.getUpdatedState({bearing, pitch});
+}
+
+// Zoom
+// @param {[Number, Number]} pos - position on screen where the current center is
+// @param {[Number, Number]} startPos - the center position at
+//   the start of the operation. Must be supplied of `zoomStart()` was not called
+// @param {Number} scale - a number between [0, 1] specifying the accumulated
+//   relative scale.
+function zoomTo(viewState, {controller, pos, startPos, scale, startViewState}) {
+  assert(scale > 0, '`scale` must be a positive number');
+
+  // Make sure we zoom around the current mouse position rather than map center
+  // startZoomLngLat: viewState._unproject(pos),
+  const startZoomLngLat =
+    controller.unproject(startPos) ||
+    controller.unproject(pos);
+
+  let {startZoom} = startViewState.zoom;
+
+  if (!Number.isFinite(startZoom)) {
+    startZoom = viewState.zoom;
+  }
+
+  // take the start lnglat and put it where the mouse is down.
+  const newZoom = startZoom + Math.log2(scale);
+  const zoomedViewport = new WebMercatorViewport(Object.assign({}, viewState, {
+    zoom: newZoom
+  }));
+  const [longitude, latitude] =
+    zoomedViewport.getLocationAtPoint({lngLat: startZoomLngLat, pos});
+
+  return viewState.getUpdatedState({
+    zoom: newZoom,  // Rely on constraints being applied
+    longitude,
+    latitude
+  });
+}
+
+function zoomIn(viewState) {
+  return viewState.getUpdatedState({
+    zoom: viewState.getZoom() + 0.2
+  });
+}
+
+function zoomOut(viewState) {
+  return viewState.getUpdatedState({
+    zoom: viewState.getZoom() - 0.2
+  });
+}
+
+function moveLeft(viewState) {
+  return viewState.getUpdatedState({
+    bearing: viewState.getBearing() - 3
+  });
+}
+
+function moveRight(viewState) {
+  return viewState.getUpdatedState({
+    bearing: viewState.getBearing() + 3
+  });
+}
+
+function moveForward(viewState) {
+  return viewState.getUpdatedState({
+    pitch: viewState.getPitch() + 3
+  });
+}
+
+function moveBackward(viewState) {
+  return viewState.getUpdatedState({
+    pitch: viewState.getPitch() - 3
+  });
+}
+
+export const EVENT_REDUCERS = {
+  // panStart,
+  pan,
+  // panEnd,
+  // rotateStart,
+  rotate,
+  // rotateEnd,
+  // zoomStart,
+  zoom: zoomTo,
+  // zoomEnd,
+  zoomIn,
+  zoomOut,
+  moveLeft,
+  moveRight,
+  moveForward,
+  moveBackward
+};
+
+// A class that handles events and updates mercator style viewport parameters
+export default class MapController extends Controller {
+
+  constructor(options) {
+    // Register map specific "ViewState + event -> ViewState" reducers
+    super(Object.assign({}, options, {reducers: EVENT_REDUCERS}));
+  }
+
+  // Override selected event-to-reducer mappings
```

Comment:
Should also override `onPan` to be press function key to rotate (align with react-map-gl behavior)

Reply:
> Should also override onPan to be press function key to rotate (align with react-map-gl behavior)

Yes. Would you be interested in doing a final pass on those things rather than back and forth in reviews?